### ISSUE TITLE
FileTarget archive issues

### ISSUE THREAD

**Issue by mckn** [NLog/NLog#646]
**The current NLog version:** 3.2.0.0
**The error message and stacktrace:** none
**The internal log:**
```
2015-03-23 15:20:05.9475 Info Loading assembly name: Logging.NLog
2015-03-23 15:20:05.9475 Debug ScanAssembly('Logging.NLog, Version=1.0.9.0, Culture=neutral, PublicKeyToken=null')
2015-03-23 15:20:05.9605 Debug Setting 'UppercaseLayoutRendererWrapper.uppercase' to 'true'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.name' to 'file'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.layout' to '${json}'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.fileName' to 'D:\Logs\Apps\${logger}\${shortdate}.log'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.archiveFileName' to 'D:\Logs\Apps\${logger}\{#}.archived'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.archiveNumbering' to 'Date'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.archiveEvery' to 'Day'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.maxArchiveFiles' to '5'
2015-03-23 15:20:05.9605 Debug Setting 'FileTarget.archiveAboveSize' to '50000'
2015-03-23 15:20:05.9765 Debug Setting 'FileTarget.archiveDateFormat' to 'yyyy-MM-dd'
2015-03-23 15:20:05.9765 Debug Wrapping target 'file' with AsyncTargetWrapper and renaming to 'file_wrapped
2015-03-23 15:20:05.9765 Info Adding target AsyncWrapper Target[file](File Target[file_wrapped])
2015-03-23 15:20:05.9765 Debug Registering target file: NLog.Targets.Wrappers.AsyncTargetWrapper
2015-03-23 15:20:05.9765 Info Configured from an XML element in C:\Development\Logging\ConsoleApplication1\bin\Debug\NLog.config...
2015-03-23 15:20:05.9765 Debug --- NLog configuration dump. ---
2015-03-23 15:20:05.9765 Debug Targets:
2015-03-23 15:20:05.9765 Info AsyncWrapper Target[file](File Target[file_wrapped])
2015-03-23 15:20:05.9765 Debug Rules:
2015-03-23 15:20:06.0015 Info logNamePattern: (:All) levels: [ ] appendTo: [ file ]
2015-03-23 15:20:06.0015 Debug --- End of NLog configuration dump ---
2015-03-23 15:20:06.0015 Info Found 15 configuration items
2015-03-23 15:20:06.0015 Debug Targets for APa by level:
2015-03-23 15:20:06.0015 Debug Trace =>
2015-03-23 15:20:06.0015 Debug Debug =>
2015-03-23 15:20:06.0015 Debug Info => file
2015-03-23 15:20:06.0015 Debug Warn => file
2015-03-23 15:20:06.0015 Debug Error => file
2015-03-23 15:20:06.0015 Debug Fatal => file
2015-03-23 15:20:06.2375 Debug Async queue is full
2015-03-23 15:20:06.2375 Debug Discarding one element from queue
2015-03-23 15:20:06.2375 Debug Async queue is full
2015-03-23 15:20:06.2375 Debug Discarding one element from queue
2015-03-23 15:20:06.2375 Debug Async queue is full
2015-03-23 15:20:06.2375 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2475 Debug Async queue is full
2015-03-23 15:20:06.2475 Debug Discarding one element from queue
2015-03-23 15:20:06.2675 Debug Async queue is full
2015-03-23 15:20:06.2675 Debug Discarding one element from queue
2015-03-23 15:20:06.2675 Debug Async queue is full
2015-03-23 15:20:06.2675 Debug Discarding one element from queue
2015-03-23 15:20:06.2675 Debug Async queue is full
2015-03-23 15:20:06.2675 Debug Discarding one element from queue
```

**The current result:**
When the archiveAboveSize is reached NLog creates one archive file that is called 2015-03-22.archived (and todays date is 2015-03-23). Then it continues to log to 2015-03-23.log but when that file is reached the archiveAboveSize it stops log. No more logs are written to file.

**The expected result:**
I am expecting NLog to write to a file called 2015-03-23.log. When a new day begins i expect NLog to archive that file as 2015-03-23.archived. If the arhciveAboveSize is reached i expects NLog to create a 2015-03-23.archived file before the end of the day. If the 2015-03-23.log once again reaches the archiveAboveSize i expect NLog to overwrite the previous 2015-03-23.archived. Same goes for when is is a new day. If there already is a 2015-03-23.archived then replace it with the last log.

**Any current work-arounds:** none

**The current config:**
```xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <extensions>
    <add assembly="Logging.NLog" />
  </extensions>
  
  <targets async="true">
    <target name="file"
           xsi:type="File"
           layout="${json}"
           fileName="D:\Logs\Apps\${logger}\${shortdate}.log"
           archiveFileName="D:\Logs\Apps\${logger}\{#}.archived"
           archiveNumbering="Date"
           archiveEvery="Day"
           maxArchiveFiles="5"
           archiveAboveSize="50000"
           archiveDateFormat="yyyy-MM-dd" />
  
  </targets>
  
  <rules>
    <logger name="*" minlevel="Info" writeTo="file" />
  </rules>
</nlog>
```

**Comment by 304NotModified** [comment]
Many thanks for the detailed bug report.

Is this a duplicate of #470?

**Comment by 304NotModified** [comment]
Possible related #643

**Comment by 304NotModified** [comment]
Possible related https://github.com/NLog/NLog/pull/383

**Comment by mckn** [comment]
Thanks for the fast feedback!

**Comment by prainho** [comment]
I think I'm having a similar problem with version 4.0.1.

I've started with this simple configuration:

      <target name="FileTrace" 
              xsi:type="File" 
              fileName="D:\logs\Trace.${shortdate}.txt" 
          layout="${longdate}|${level:uppercase=true}|${machinename}|${logger}|${message}|${exception:format=message,type,stacktrace,method:maxInnerExceptionLevel=5:innerFormat=shortType,message,method}"/>

Then I've added a configuration do archive on every 25mb

      <target name="FileTrace" 
              xsi:type="File" 
              fileName="D:\logs\Trace.${shortdate}.txt" 
              layout="${longdate}|${level:uppercase=true}|${machinename}|${logger}|${message}|${exception:format=message,type,stacktrace,method:maxInnerExceptionLevel=5:innerFormat=shortType,message,method}"
              archiveFileName="D:\logs\archives\Trace.{#}.zip"
              archiveAboveSize="26214400"
              archiveNumbering="Date"
              archiveDateFormat="yyyyMMdd.HHmm"
              maxArchiveFiles="100"
              enableArchiveFileCompression="true"/>

With this configuration I've found that if in the end of the day the file is less than 25mb and the file is not archived and a new log file is created.

Than I've added the option to archive every day:

      <target name="FileTrace" 
              xsi:type="File" 
              fileName="D:\logs\Trace.${shortdate}.txt" 
              layout="${longdate}|${level:uppercase=true}|${machinename}|${logger}|${message}|${exception:format=message,type,stacktrace,method:maxInnerExceptionLevel=5:innerFormat=shortType,message,method}"
              archiveFileName="D:\logs\archives\Trace.{#}.zip"
              archiveAboveSize="26214400"
              archiveEvery="Day"
              archiveNumbering="Date"
              archiveDateFormat="yyyyMMdd.HHmm"
              maxArchiveFiles="100"
              enableArchiveFileCompression="true"/>

And with this configuration something strange happen.

1º the zip with file above 25mb wasn't created (So archiveAboveSize didn't work)
2º when the file was above 25mb a new log file was created and I lost 25mb of logs
3º the archive every day didn't work either.

Now the expected behavior:
1º create a new zip file for every 25mb of logs
2º create a zip file every day

Now since this is a bug and it's expected to be fixed in version 4.1 what archive configuration should I use while this problem is begin fixed.

It would could also be nice to have a configuration similar to this:
* archiveEvery="Half-Day" - archive the file every 12 hours

**Comment by prainho** [comment]
Ok, I've done a new test this time with this configuration

<target name="FileTrace" 
          xsi:type="File" 
          fileName="D:\logs\Trace.${shortdate}.txt" 
          layout="${longdate}|${level:uppercase=true}|${machinename}|${logger}|${message}|${exception:format=message,type,stacktrace,method:maxInnerExceptionLevel=5:innerFormat=shortType,message,method}"
          archiveFileName="D:\logs\archives\Trace.{#}.zip"
          archiveEvery="Day"
          archiveNumbering="Date"
          archiveDateFormat="yyyyMMdd.HHmm"
          maxArchiveFiles="100"
          enableArchiveFileCompression="true"/>

The last trace in file Trace.2015-07-09.txt is at 2015-07-09 18:40:59.2925
The fisrt trace in file Trace.2015-07-10.txt is at 2015-07-10 10:30:55.5020

And no archive was made. So it seems to me that archive every day is not working.

**Comment by prainho** [comment]
Ok, I've done a new test this time with this configuration

xsi:type="File" 
fileName="D:\logs\Trace.${shortdate}.txt" 
layout="${longdate}|${level:uppercase=true}|${machinename}|${logger}|${message}|${exception:format=message,type,stacktrace,method:maxInnerExceptionLevel=5:innerFormat=shortType,message,method}"
archiveFileName="D:\logs\archives\Trace.{#}.zip"
archiveEvery="Hour"
archiveNumbering="Date"
archiveDateFormat="yyyyMMdd.HHmm"
maxArchiveFiles="100"
enableArchiveFileCompression="true"/>

And worked.

**Comment by prainho** [comment]
Ok, I've done a new test this time with this configuration

xsi:type="File" 
fileName="D:\logs\Trace.${shortdate}.txt" 
layout="${longdate}|${level:uppercase=true}|${machinename}|${logger}|${message}|${exception:format=message,type,stacktrace,method:maxInnerExceptionLevel=5:innerFormat=shortType,message,method}"
archiveFileName="D:\logs\archives\Trace.{#}.zip"
archiveEvery="Hour"
archiveNumbering="Date"
archiveDateFormat="yyyyMMdd.HHmm"
maxArchiveFiles="100"
enableArchiveFileCompression="true"
ArchiveOldFileOnStartup="true"/>

And when the site started the traces in Trace.2015-07-10.txt were archived.
But the previous days traces in Trace.2015-07-9.txt, Trace.2015-07-8.txt,etc... were not archived. It would be great to archive old trace files with the same pattern that for some reason weren't archived yet.

**Comment by 304NotModified** [comment]
hi @prainho 

Thanks for the test cases. It's true that this issue isn't solved yet. 

Can you summarize when it works and when don't? That would be really helpful in fixing this case.

**Comment by 304NotModified** [comment]
This seems the (partly) the same bug as https://github.com/NLog/NLog/issues/569

**Comment by 304NotModified** [comment]
same issue as #569. 


The combinations: 


* archiveEvery="Hour"
*  archiveNumbering="Date"


and 


*  archiveAboveSize="26214400"
*  archiveNumbering="Date"

are **not** supported. We don't merge archive files. Please use archiveNumbering="DateAndSequence".

see https://github.com/NLog/NLog/wiki/File-target#archival-options

I added a warning at this page.
